### PR TITLE
[NVPTX] Add conversion intrinsics from/to fp8 types (e4m3, e5m2)

### PR DESCRIPTION
**Pull request by sergey-kozub**
PTX ISA 8.1 supports FP8 conversions:
https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cvt

This PR adds the support for:

- cvt.rn.satfinite{.relu}.f8x2type.f32 d, a, b;
- cvt.rn.satfinite{.relu}.f8x2type.f16x2 d, a;
- cvt.rn.{.relu}.f16x2.f8x2type d, a;

where .f8x2type = { .e4m3x2, .e5m2x2 };

### CODE REVIEW

File: clang/test/CodeGen/builtins-nvptx.c
Comment (Artem-B) [inline clang/test/CodeGen/builtins-nvptx.c:978]: Nit: 1->1.0f. While implicit conversion works, it looks a bit odd when the check line matches `float 1.000000e+00`

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:747]: It would be useful to add a comment that the output i32 is actually <2 x f16> as `x2` does not give enough of a hint. Or, perhaps rename it to `_f16x2`

**Review by Artem-B — APPROVED** [review]
LGTM with a couple of nits.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (sergey-kozub) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:747]: This one is confusing, "x2" is only there only because (for some reason) I cannot specify an empty suffix.
All the classes in this file use the `CVT_<to>_<from>` pattern. I need to define two classes: CVT_f16x2_e4m3x2 and CVT_f16x2_e5m2.

You're suggesting to rename to `_f16x2` but I guess "multiclass" only supports appending the suffixes, whereas I need "f16x2" to appear in the middle of the class name (in order to conform to the common naming pattern).

Also, "outs" is Int32Regs, and I cannot specify <2xf16> here. This is in line with the other class definitions in this file.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:747]: I think you should be able to specify explicit name by using the magic `NAME` value to construct the name the way you want to. E.g: https://github.com/llvm/llvm-project/blob/101acff2e5df21378b5764660afc0ec84ce6a1a6/llvm/include/llvm/IR/IntrinsicsNVVM.td#L1450

> Also, "outs" is Int32Regs, and I cannot specify <2xf16> here. This is in line with the other class definitions in this file.

This part is fine, as Int32 is the register type for the `2xf16`. I was only concerned about the names.

Comment (sergey-kozub) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:747]: Neat, wasn't aware of the NAME keyword. Updated the PR.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:745]: We do not need a multiclass here. It could be just class.

Comment (sergey-kozub) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:745]: Replaced with "class"

**Review by Artem-B — APPROVED** [review]

**Comment by sergey-kozub** [comment]
What do I need to push this through? I'm still seeing "1 workflow awaiting approval" message.
Should I add some other reviewer(s) who are authorized to run workflows?

**Comment by Artem-B** [comment]
Looks like that it was clang-format check github was waiting on an approaval for. I've just clicked that button, let's see what it brings.
The patch is good to go otherwise, IMO.

**Comment by justinfargnoli** [comment]
> How do I actually upstream it? 

Here is the documentation on how to [obtain commit access](https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access). It's a relatively quick process. 

However, if you don't want to go through that process, you can just ask for it to be committed, and someone will do it on your behalf.

**Comment by sergey-kozub** [comment]
Thanks for pushing this, I'll continue by adding the support to XLA to use these instructions.

**Pull request merged**
